errors.go

```diff
@@ -373,3 +376,24 @@ var (
 
 	ErrStateVersionUploadNotSupported = errors.New("upload not supported by this version of Terraform Enterprise")
 )
+
+type DecodeHTTPError struct {
```

Comment:
This looks like a useful tool, but it doesn't help for types that are already response errors that can be decoded, like the ones returned by checkResponseCode.

You could add a function that can decode the contents of a variety of errors, starting with ErrResourceNotFound, that can be built with a response body and work in tandem with checkResponseCode.

It's important that the existing message does not change for ErrResourceNotFound. The additional decoded details should be available from a different field and not from the Error() method.

```
type ErrResourceNotFound {
  Details []string
}

func (ErrResourceNotFound) Error() string {
  return "resource not found" // <-- this is the important compatibility part
}

func NewErrorFromResponseCode(response *http.Response) error {
  decodedDetails, err := decodeErrorPayload(response)
  if err != nil {
    // There aren't any error details so this error isn't a 'jsonapi error'
    return fmt.Errorf("could not jsonapi error from this response: %w", err)
  }
  
  switch response.StatusCode {
  case 404:
    return ErrResourceNotFound{Details: decodedDetails}
  // etc
  }
  
  return nil
}
```

Reply:
Also don't forget unit tests

Reply:
Unit tests will need to be updated because the error will be wrapped with this solution.
Likely tests that check for equality:
```
assert.Equal(t, ErrResourceNotFound, err)
```